Crowdin CSV & Translation Context

nucleogenesis opened the pull request:
<!--
 1. Following guidance below, replace …'s with your own words
 2. After saving the PR, tick of completed checklist items
 3. Skip checklist items that are not applicable or not necessary
 4. Delete instruction/comment blocks
-->

### Summary
<!--
 * description of the change
 * manual verification steps performed
 * screenshots if the PR affects the UI
-->

fixes #5582

This PR covers several related updates to the overall release and translation process.

__The New Process Summary__

1. During Development (Prior to String Freeze)
    - Developers can add context for a string directly in the code in any place where strings are defined. This includes calls to `createTranslator` and the assignment to `$trs` in a Vue component.
    - Any updates made to this development branch on Crowdin are liable to be overwritten by whatever exists in the code.

2. Enter String Freeze
    - Once we have entered string freeze, the release manager will run `make i18n-upload branch={branch}` which parses the code into CSV files which are then uploaded to the new branch on Crowdin. __This process includes pushing the context that was added in the code directly to Crowdin. That context may then later be modified only on the Crowdin UI__
    - The translation process proceeds as normal from here in all ways including the pre-translation, manual approval, addition of context, etc.

3. Ready to Release 
    - The release manager will run `make i18n-download branch={branch}`, which will do the following:
      > 1.  Download all of the CSVs from Crowdin, including __all__ translations for every language and all context.
      > 2. Using the data in those CSVs, create the exact same `whatever_module-messages.json` files that we previously were downloading from Crowdin. These files are used by our `i18n.js` to load all translations for each language.
      > 3. Further making use of the data in those CSVs, it will update the entire codebase to reflect all of the contexts that were added and approved on Crowdin so that it is represented in the codebase in the proper format.
    - The release manager then will have to stage and commit the changed Vue & JS files where the context was added or otherwise modified across the codebase.

4. New Release Target - Rinse & Repeat
    - Context is now persisted in our codebase and may be changed throughout the development cycle. It will never be lost between Crowdin translation releases, ideally saving substantial time for our i18n team and translators.

Example of context being added to code.

```javascript
$trs: {
    // Expected when there is context. If we ever download context, it will be applied as follows.
    usernameLabel: {
        message: "Username",
        context: "A label for the 'username' field on forms.",
    },
    // Acceptable, will not cause an error but isn't ideally formatted if there is no context. In this case, the transfer process will rewrite this as a string without the object definition.
    passwordLabel: {
        message: "Password",
    },
   // Also acceptable. During the `i18n-download` process, this will be converted to the following:
   passwordConfirmationLabel: "Confirm password",
}
```

### Reviewer guidance
<!--
 * how can a reviewer test these changes?
 * are there any risky areas that deserve extra testing
-->

# How to Test This PR

This is a rough end-to-end outline for manual testing of this PR.

## Before String Freeze

Imagine that the code base is in flux prior to our cut-off for an imminent release. We will want to add context to parts of the code and ensure that it is sent to a new Crowdin branch for our hypothetical imminent release.

### Update several files where strings are defined to include context strings.

Add context to the code. _Take note of which strings you've updated to include context so that you can easily find them on Crowdin and verify that context was indeed transfered._

- Update strings in one or more of the Common*Strings.js files.
- Update strings defined elsewhere in *.js files - you can search the code base for places where `createTranslator` is called and add context to some of the strings defined there.
- Update strings in any number of *.vue files where an `$trs` property is defined on the Vue component.

**Adding Context**

```javascript
$trs: {
	// Change it from this:
	fooLabel: 'Foo',
	// To this:
	fooLabel: {
		message: 'Foo',
		context: 'Bar baz.',
	},
}
```

## Now it is time for String Freeze

### Upload to Crowdin

- Run `make i18n-upload branch={make-up-a-unique-branch-name}`. This will process the whole code base, including the context you just added, and will upload it to your newly defined branch.
- Go to Crowdin, find the translations for your `made-up-branch-name`, and review your notes from the last step. Check that your context was added for all of the strings that you changed in the code base.

### Modify Context on Crowdin

In our hypothetical release process, we are now allowing our translators to work their magic. We will add and modify the context on Crowdin directly. _It will help to take notes of which strings you've changed so you can review them later._

- Go to your branch on Crowdin - **Your Branch!!**.
- Go to strings that have no context and add some context. Save and approve the string.
- Go to strings that have context which was successfully uploaded previously and change the context text. Save and approve that string.

## Time to Prepare the Release!

### Download From Crowdin

- Run `make i18n-download branch={your-made-up-branch-name}`
- Run `make i18n-transfer-context`
- Now, review the changes that have been made to the code and you ought to see all of the changes you just made in Crowdin reflected directly in the code base.

## Another Release Cycle

In order to be sure that context persists across releases, basically you should repeat all of the above. Modify context that you've added originally, add context to more strings, etc.

## Known Issues *Help Very Much Welcome!*

### Template comment formatting & destruction

On files where comments are defined outside of a <template> <script> or <style> block, they will be removed by the AST reading/writing process. For this issue, it seems that we will just need to be mindful of this as we proceed in future commits / PRs. When a comment is found to be outside of the blocks, it should  be moved into the most relevant block.

Additionally, some comments may not be output with proper formatting.

Frankly, I have no idea why this is the case or what can be done to resolve it. __Nothing happens to the `<template>` content during this process__. It is read and re-written between `<template>` tags and that is all.

### Need to re-run yarn lint-frontend:format again

The yarn command that runs the context transfer already chains a `&& yarn lint-frontend:format` but in some cases, I have had to re-run it to take care of some lingering formatting issues. 

### Enforced order of Vue blocks.

When rewriting updated files, no matter what the order is of the <template> <script> and <style> tags, they will always be rewritten in that order.

I have only noticed this as an issue in one file, `kolibri/core/assets/src/views/KTooltip/Popper.vue`. If this is not acceptable, I suspect it can be resolved during the re-writing process, but I passed on putting the effort since that is the "proper" flow of a Vue file and the rare case that this would ever come up.

Note that a file is not re-written if nothing has changed.

### Issues chaining context transfer with download

Since the `make i18n-context-transfer` runs `yarn transfercontext` which in turn runs `yarn lint-frontend:format` - if we chain that linting into the `make` process and the linter returns a failure (which it seems to do no matter what) - then the `make` process claims to have also failed because one of the commands returned as failing.

So - currently we have to run each of the commands separately to avoid this issue. We could instead _not_ chain the `yarn lint-frontend:format` and require to run it manually. I'm definitely open to suggestions.

### References
<!--
 * references to related issues and PRs
 * links to mockups or specs for new features
 * links to the diffs for any dependency updates, e.g. in iceqube or the perseus plugin
-->

[Strings Refactor on Notion](https://www.notion.so/learningequality/String-refactor-e79dd8c431ef4a1084142ebf5e81ccab)

----

### Contributor Checklist


PR process:

- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

Testing:

- [x] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

### Reviewer Checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

nucleogenesis committed:
yarn makemessages output to CSV

nucleogenesis committed:
prepare for csv upload to crowdin

rtibbles commented:
The only other way I could think to do it would be with comment parsing, but I think this is more explicit, and less error prone.

As an example this is what the other way could look like:
```
$trs: {
    helloWorld: 'Hello world',
   /* i18n-context-next 
   *  Excitedly greeting the user with a friendly welcome the first time they log in.
   */
    welcomeLabel: 'Welcome!',
}
```

indirectlylit commented:
Copying some comments from a slack thread here:

> I think that proposal above is a very good idea, but there might be pros and cons related to other options. Two downsides come to mind:
>
> * context is more likely to be edited from the crowdin UI by folks like Radina than it’s edited by developers directly in code
> * context is often going to be multiple sentences long including endlines, risking bloat to vue files 
> 
> One additional consideration is that in addition to textual context, we also want eventually to be able to associate screenshot(s) with strings.  Ideally, we’d store the screenshot in git, along with our context strings and translations. Then we’d update `crowdin.py` to automatically upload and download screenshots using the [V2 crowdin API](https://support.crowdin.com/api/v2/#operation/api.projects.screenshots.createOne). this would be a new step in the process alongside the existing crowdin automation
> 
> Here are a couple others strategies that I can think of…
> 
> 1. Currently we don’t track English .json files in Git. We could potentially start tracking the new csv files in git, and context would be stored in there
> 2. We could potentially create a new `locale-context` directory (or similar) which contains both screenshots and text-based context. There would be e.g. a new `context.json` file like
>      `{ "[id.namespace]": "[context]", ... }`
>

nucleogenesis committed:
Parse downloaded CSVs into namespaced JSON

nucleogenesis committed:
Move CSV-to-JSON process to own command, makefile

nucleogenesis committed:
Allow object definition for string/context

nucleogenesis committed:
Process context trransfer

nucleogenesis committed:
Transfer context, yarn format, make command

nucleogenesis committed:
Include context transfer in Crowdin download process

nucleogenesis committed:
Make functions slightly more testable

nucleogenesis committed:
Imrroved AST generation, remove from makefile

nucleogenesis committed:
Test suite for context transfer

nucleogenesis committed:
Python formatting

indirectlylit commented on packages/kolibri-tools/lib/i18nSyncContext.js:
```diff
@@ -0,0 +1,343 @@
+// Processing Functions //
+// -------------------- //
+
+const processVueFiles = (files, definitions) => {
```
is there a particular reason we're defining functions in here as 

```js
const funcName = (args) => {}
```

instead of 

```js
function funcName(args) {}
```

if we have a choice, I think the latter is much more clear

nucleogenesis replied:
@indirectlylit sorry about that - honestly it is just a habit. I like the arrow function format. I've updated all functions defined at the root level with standard JS function declarations.

However, for callbacks passed to various methods (ie, `[1,2,3].map(i => ...)`) I have kept the functions defined as arrow functions.

nucleogenesis committed:
Ensure CSVs parse to JSON without headers

nucleogenesis committed:
Remove arrow functions, handle TemplateLiterals

nucleogenesis committed:
Lint, remove upload-translation from make

nucleogenesis committed:
More linting & cleanup

nucleogenesis committed:
Python format

nucleogenesis committed:
Ignore i18n linting in context fixtures

jonboiser commented:
We call the messages 'strings' as developers, but maybe we should use `message` instead of `string` in the new `$trs` values to match the i18n API and terminology?

nucleogenesis committed:
Only modify between <script> tags in Vue files

nucleogenesis committed:
Skip node_modules in context sync,  lint

nucleogenesis committed:
Use the key `message` instead of `string`

nucleogenesis committed:
Handle object definitions of messages/context in i18n.js

nucleogenesis committed:
Clean up var names in extraction, fix missing strings

nucleogenesis committed:
Include yarn.lock

nucleogenesis committed:
Fix bug writing every file no matter what

radinamatic commented:
> Fix bug writing every file no matter what 

O'yeah, I was wondering about that `Kolibri wrote a file!` after every line in the Terminal 😂

nucleogenesis committed:
Remove pathlib for Python2 comaptibility

nucleogenesis committed:
Print single-quotes, improved script injection

nucleogenesis committed:
<script> injection fix

nucleogenesis committed:
Append Identifier to context on upload, remove on ctx tx

radinamatic commented:
Hey @nucleogenesis! 
After all the testing and debugging we did last week, fixing the initial issues, I did manage to do one full circle of edit-locally-upload-edit-on-crowdin-download, with the test-runs on the dev server in between to see if everything builds smoothly. Awesome job so far, we are so close! 👍 

I just re-tried with your latest changes after our call on Friday, and run into some issues: `make i18n-upload branch=...` finished without problems, and I could see my second test branch on Crowdin, but the devserver displayed some webpack issues related to Perseus at this point.

![2019-09-16_1-41-51](https://user-images.githubusercontent.com/1457929/64929596-476b0c00-d828-11e9-8e34-1e3fe33e71b5.png)

However, I proceeded to make changes in the context of the uploaded CSVs on Crowdin, but after running `make i18n-download branch=...` I got this error which is different from the one we solved with `yarn lint-frontend:format` the last time, so my guess is that something else has gone awry.

```
(kolibri2) didivm@didivm-VirtualBox:~/LE/kolibri$ make i18n-download branch=radina-test2
python build_tools/i18n/crowdin.py rebuild radina-test2
INFO: Crowdin: rebuilding 'radina-test2'. This could take a while...
INFO: Crowdin: rebuild succeeded! built
python build_tools/i18n/crowdin.py download radina-test2
INFO: Crowdin: downloading 'radina-test2'...
INFO: 	Extracting ar to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting bg to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting bn to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting es-ES to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting la to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting fa to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting fr to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting fv to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting gu-IN to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting hi to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting mr to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting my to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting ny to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting pt-BR to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting sw-TZ to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting te to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting ur-PK to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting vi to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting yo to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting ach to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
Traceback (most recent call last):
  File "build_tools/i18n/crowdin.py", line 656, in <module>
    main()
  File "build_tools/i18n/crowdin.py", line 637, in main
    command_download(args.branch)
  File "build_tools/i18n/crowdin.py", line 407, in command_download
    _format_json_files()  # clean them up to make git diffs more meaningful
  File "build_tools/i18n/crowdin.py", line 301, in _format_json_files
    json_data = _locale_data_from_csv(f, locale_path)
  File "build_tools/i18n/crowdin.py", line 334, in _locale_data_from_csv
    headers = csv_reader.__next__()
AttributeError: '_csv.reader' object has no attribute '__next__'
Makefile:210: recipe for target 'i18n-download' failed
make: *** [i18n-download] Error 1
(kolibri2) didivm@didivm-VirtualBox:~/LE/kolibri$
```

On another hand, regarding the position of the string identifier we discussed the last time, once the CSV is loaded into Crowdin translator UI, identifier is still visible as the first item in the context box, whether there was context in the code (as in the string below) or not.

![2019-09-16_1-12-28](https://user-images.githubusercontent.com/1457929/64929694-68802c80-d829-11e9-8fd7-7233539dc7ed.png)

As I mentioned, my concern is that if the identifier is the first thing in the context box (as it reads like code, much more than actual _context_ info), translators may disregard the rest that comes below. The ideal situation would be that when there is context specified in the code, once CSV is uploaded it should display as the first thing in the context box on Crowdin (see below), and the `-- Do not edit this line or below! --` (sentence-case here, please, not the title-case 😉) and the identifier itself appear below the context, **with at least one empty line** in between.

![2019-09-16_1-19-59](https://user-images.githubusercontent.com/1457929/64929731-0bd14180-d82a-11e9-836e-68df4de129fa.png)

Let me know if this is doable!

nucleogenesis committed:
Modify context line addition

indirectlylit commented:
exciting progress

nucleogenesis commented:
Just noting here for posterity.

Per Radina's last comment - regarding appending the `Identifier` to the actual context string.

Unfortunately, this is not doable due to limitations in Crowdin. I contacted Crowdin's support (they were awesome and quick!) and was informed that we would not be able to remove the `Identifier` that Crowdin always imposes on the top of the context.

So we have opted for a different process. We will append the context to the Identifier - so on Crowdin, you will see:

```
Component.key

-- CONTEXT --
The context defined in the code somewhere
```

The code now automatically adds the `-- CONTEXT -- ` part when it uploads and strips it out when it downloads so - it's just some background magic that does not affect the use of the new context-in-code system in any way.

nucleogenesis committed:
Refactor upload/download conversion process

nucleogenesis committed:
Keep downloaded CSVs out of git

nucleogenesis committed:
Backend lint

radinamatic commented:
Upload goes well, download not so much, same `AttributeError: '_csv.reader' object has no attribute '__next__'` error as before :disappointed: 

```
(kolibri) didivm@didivm-VirtualBox:~/LE/kolibri$ make i18n-upload branch=radina-test4
yarn run makemessages
yarn run v1.17.3
$ kolibri-tools build i18n --file ./build_tools/build_plugins.txt
Gathering relevant modules from ['kolibri.core', 'kolibri.plugins.*']
WARNING:root:No C Extensions available for this platform.
///...///
INFO: Crowdin: upload succeeded!
(kolibri) didivm@didivm-VirtualBox:~/LE/kolibri$ make i18n-download branch=radina-test4
python build_tools/i18n/crowdin.py rebuild radina-test4
INFO: Crowdin: rebuilding 'radina-test4'. This could take a while...
INFO: Crowdin: rebuild succeeded! built
python build_tools/i18n/crowdin.py download radina-test4
INFO: Crowdin: downloading 'radina-test4'...
INFO: 	Extracting ar to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting bg to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting bn to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting es-ES to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting la to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting fa to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting fr to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting fv to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting gu-IN to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting hi to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting mr to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting my to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting ny to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting pt-BR to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting sw-TZ to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting te to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting ur-PK to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting vi to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting yo to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
INFO: 	Extracting ach to /home/didivm/LE/kolibri/kolibri/locale/CSV_FILES
Traceback (most recent call last):
  File "build_tools/i18n/crowdin.py", line 629, in <module>
    main()
  File "build_tools/i18n/crowdin.py", line 610, in main
    command_download(args.branch)
  File "build_tools/i18n/crowdin.py", line 382, in command_download
    _format_json_files()  # clean them up to make git diffs more meaningful
  File "build_tools/i18n/crowdin.py", line 307, in _format_json_files
    data = _locale_data_from_csv(f)
  File "build_tools/i18n/crowdin.py", line 316, in _locale_data_from_csv
    csv_reader.__next__()  # Remove the headers
AttributeError: '_csv.reader' object has no attribute '__next__'
Makefile:210: recipe for target 'i18n-download' failed
make: *** [i18n-download] Error 1
(kolibri) didivm@didivm-VirtualBox:~/LE/kolibri$ 
```

nucleogenesis committed:
Make compatible for Pythons 2 and 3

nucleogenesis committed:
Fix key in JSON conversion

nucleogenesis committed:
language_globals coreMessages key was wrong

nucleogenesis committed:
Remove unnecessary JSON processing

nucleogenesis committed:
Account for JSON unicode encoding in dump

radinamatic commented:
I think we finally got there! :tada: :+1: :star2: 

I've gone through the steps several time on a couple of test branches, changed and added context locally, uploaded, changed/added on Crowdin, downloaded, transferred and checked that context displays, retested that devserver is running, and then did all again from the start. A part from the known issue that linter errors out and you need to rerun the `yarn lint-frontend:format`, I encountered no other issues.

Thank you for the excellent job, and for pushing through this final stretch. This is going to improve our work of supporting translators much easier, and I can't wait to start using it!

radinamatic reviewed (approved):
Such a beauty... :trophy:

rtibbles reviewed (changes requested):
Still want to revert the `JSON.parse` change here, and use the Py2/3 sensitive `ensure_ascii=False` for dumping the JSON into the template.

nucleogenesis committed:
Pass properly escaped JSON in plugin hook

nucleogenesis committed:
Include `ensure_ascii` in frontned_message_tag json.dumps

nucleogenesis commented:
> Still want to revert the `JSON.parse` change here, and use the Py2/3 sensitive `ensure_ascii=False` for dumping the JSON into the template.

I've added `ensure_ascii=False` to `plugin_data_tag()` and `frontend_message_tag()`.

Additionally, I've added a `replace()` to escape double-quotes in the `JSON.parse()` in the `plugin_data_tag()` method.

nucleogenesis committed:
JSON.parse properly escaped JSON in frontend_message_tag

nucleogenesis committed:
linting the backend
